torch/_functorch/aot_autograd.py

```diff
@@ -3029,27 +3084,43 @@ def runtime_wrapper(*args):
                 )
 
         num_mutated_inps = runtime_metadata.num_mutated_inputs
+        num_mutated_runtime_inps = runtime_metadata.num_mutated_inp_runtime_indices
         num_metadata_mutated_inps = runtime_metadata.num_mutated_metadata_inputs
         num_intermediate_bases = runtime_metadata.num_intermediate_bases
 
         if keep_input_mutations:
-            assert (
-                len(all_outs)
-                == num_metadata_mutated_inps + runtime_metadata.num_outputs + num_intermediate_bases
-            )
-            assert (
-                len(runtime_metadata.mutated_inp_runtime_indices) == num_metadata_mutated_inps
-            )
+            if not trace_joint:
+                assert (
+                    len(all_outs)
+                    == num_metadata_mutated_inps + runtime_metadata.num_outputs + num_intermediate_bases
+                )
+                assert (
+                    runtime_metadata.num_mutated_inp_runtime_indices == num_metadata_mutated_inps
+                )
+            else:
+                num_graph_handled = runtime_metadata.num_mutated_graph_handled_indices
+                # We return graph handled inputs in CompiledFunction.forward as we marked them with mark_dirty
+                if num_graph_handled > 0:
+                    all_outs = all_outs[:-num_graph_handled]
```

Comment:
Ah right. Can you add a comment here mentioning that `autograd.Function` **requires** us to return the mutated inputs as extra outputs to the `autograd.Function.forward`? I think that's the real reason we're forced to do that here.